### PR TITLE
Primitive related cleanups

### PR DESCRIPTION
**Pull request by muellmusik**
This branch contains a number of cleanups and fixes related to primitives.

- Several primitives were returning values other than errNone, but lacked calls to primitiveFailed in the SC method defs, so would fail silently rather than throw the expected errors
- Speech primitives did not check for wrong types where appropriate
- Some old SC.app cruft which required (now) unbound primitives was removed
- I updated the Writing Primitives help with more info about how to write sclang methods which call primitives in order to throw any errors or have fallback SC implementations.

### CODE REVIEW

**Comment by telephon** [comment]
As this fixes many bugs, I suggest merging it into 3.7 – yes?

**Comment by muellmusik** [comment]
Yes, I think so. Let's give a little chance for comment though. In general, I can't see any reason why the errors should be suppressed, but perhaps there was some idea behind it in one case or another.

File: lang/LangPrimSource/SC_Speech.mm [hunk 1]
```diff
@@ -108,8 +108,9 @@ int prInitSpeech(struct VMGlobals *g, int numArgsPushed){
 	OSErr theErr = noErr;
 	//PyrSlot *a = g->sp-1;
 	PyrSlot *b = g->sp;
-    int chan;
-    slotIntVal(b, &chan);
+    int chan, err;
+    err = slotIntVal(b, &chan);
```
Comment (timblechmann) [inline lang/LangPrimSource/SC_Speech.mm]: btw, it is always a good idea to declare the variables exactly where they are used. having to declare all local variables at the beginning of the function is a pre-c99 requirement for c, but sc is using c++.
especially, it doesn't make the code exactly readable

**Comment by muellmusik** [comment]
@timblechmann Did you see my post on sc-dev? (Not sure if you're following these days)

https://www.listarc.bham.ac.uk/lists/sc-dev/msg54659.html

prStringCompare also seems wrong, but I'm not sure if there was perhaps some reason to suppress errors. I can add a commit to deal with this if nobody knows why it's the way it is.

**Comment by timblechmann** [comment]
@muellmusik i'm hardly following sc-dev these days ...

had a quick look:
* if a primitive fails, the class library should return `this.primitiveFailed` after the invocation of the primitive. `String-compare` doesn't do this. of course this is only necessary when the primitive actually might fail (not returning `errNone`) ... this looks good.
* there are different 'reasonable' ways to define the semantics of `String-compare`. (a) it could only support `String`s, (b) it could support `Symbols` as well or (c) it could support any type that can be converted to a `String`. i guess (b) would follow the 'principle of least surprise'.
* `prStringCompare` should probably return `errWrongType` instead of silently failing and returning `nil`.

that's all related to supercollider's optimistic software architecture and the lack of error handling. it's great to see that you're doing some cleanups on this!

**Comment by muellmusik** [comment]
> if a primitive fails, the class library should return this.primitiveFailed after the invocation of the primitive. String-compare doesn't do this. of course this is only necessary when the primitive actually might fail (not returning errNone) ... this looks good.

Yes, I added a note about this and return values in the Writing Primitives help.

> prStringCompare should probably return errWrongType instead of silently failing and returning nil.

Yes, with the current functionality it seemed so to me as well.

> It's great to see that you're doing some cleanups on this!

Cheers. I think I've got most of the ones that should have been and weren't. At least with the doc having been beefed up people will have a reasonable reference for this, GC stuff, etc. Actually if you have a moment to read over that and add anything you think is missing that would be great.

BTW, would be nice to get back to GC/smart pointer/improvements thing at some point. Another grey count error recently appeared. :-/

**Comment by timblechmann** [comment]
> Cheers. I think I've got most of the ones that should have been and
> weren't. At least with the doc having been beefed up people will have a
> reasonable reference for this, GC stuff, etc. Actually if you have a
> moment to read over that and add anything you think is missing that
> would be great.
> 
> BTW, would be nice to get back to GC/smart pointer/improvements thing at
> some point. Another grey count error recently appeared. :-/

that's more a 2-days than a 2 hours project as new allocations would
invalidate the smart pointer state ...
one could keep a set of temporary GC roots, but again this would require
some non-trivial API changes to all primitives that perform allocations

**Comment by muellmusik** [comment]
After consulting a bit, I've changed 3 string primitives to throw wrong type errors. I can't imagine why someone would have relied on the existing behaviour (fail silently and set result to nil), but review carefully.

**Comment by telephon** [comment]
As long as ```Symbol``` doesn't rely on it, it's ok:

```\hallo < \world```

should return true

**Comment by muellmusik** [comment]
> As long as Symbol doesn't rely on it

Doesn't seem to.

**Comment by telephon** [comment]
Yes, I also didn't expect it, but good to be sure. The code looks good. If the primitive is only ever used in ```String``` then we shouldn't expect anyone to use the primitive from somewhere else.

**Comment by muellmusik** [comment]
@telephon there seems to be an issue with these String changes in NdefMixer(s);
```
ERROR: Primitive '_StringCompare' failed.
Wrong type.
RECEIVER:
Instance of String {    (0x1155ff8d8, gc=A4, fmt=07, flg=00, set=02)
  indexed slots [21]
      0 : N
      1 : d
      2 : e
      3 : f
      4 : .
      5 : a
      6 : l
      7 : l
      8 : [
      9 : '
     10 : l
     11 : o
     12 : c
     13 : a
     14 : l
     15 : h
     16 : o
     17 : s
     18 : t
     19 : '
     20 : ]
}
PATH: /Users/scottw/Public/Corfu/massradiocodingfilestreamCorfu.scd
CALL STACK:
	MethodError:reportError   0x1120657f8
		arg this = <instance of PrimitiveFailedError>
	Nil:handleError   0x1123f3108
		arg this = nil
		arg error = <instance of PrimitiveFailedError>
	Thread:handleError   0x1153272a8
		arg this = <instance of Thread>
		arg error = <instance of PrimitiveFailedError>
	Object:throw   0x115326fe8
		arg this = <instance of PrimitiveFailedError>
	Object:primitiveFailed   0x112065958
		arg this = "Ndef.all['localhost']"
	String:compare   0x1125a9a58
		arg this = "Ndef.all['localhost']"
		arg aString = nil
		arg ignoreCase = false
	String:!=   0x112066668
		arg this = "Ndef.all['localhost']"
		arg aString = nil
	ProxyMixer:checkUpdate   0x115224c28
		arg this = <instance of NdefMixer>
		var newState = <instance of Event>
		var arNames = nil
		var prevArNames = nil
		var fullSize = nil
		var krNames = nil
		var prevKrNames = nil
		var editName = nil
		var prevEditName = nil
	JITGui:object_   0x112168c08
		arg this = <instance of NdefMixer>
		arg obj = <instance of ProxySpace>
	NdefMixer:object_   0x1153208c8
		arg this = <instance of NdefMixer>
		arg obj = <instance of Server>
	Interpreter:interpretPrintCmdLine   0x1163a8848
		arg this = <instance of Interpreter>
		var res = nil
		var func = <instance of Function>
		var code = "NdefMixer(s)"
		var doc = nil
		var ideClass = <instance of Meta_ScIDE>
	Process:interpretPrintCmdLine   0x11632d4f8
		arg this = <instance of Main>
^^ The preceding error dump is for ERROR: Primitive '_StringCompare' failed.
Wrong type.
RECEIVER: Ndef.all['localhost']


ERROR: Primitive '_StringCompare' failed.
Wrong type.
RECEIVER:
Instance of String {    (0x1155a1878, gc=A8, fmt=07, flg=00, set=02)
  indexed slots [21]
      0 : N
      1 : d
      2 : e
      3 : f
      4 : .
      5 : a
      6 : l
      7 : l
      8 : [
      9 : '
     10 : l
     11 : o
     12 : c
     13 : a
     14 : l
     15 : h
     16 : o
     17 : s
     18 : t
     19 : '
     20 : ]
}
PATH: /Users/scottw/Public/Corfu/massradiocodingfilestreamCorfu.scd

PROTECTED CALL STACK:
	Meta_MethodError:new	0x11338f280
		arg this = PrimitiveFailedError
		arg what = Wrong type.
		arg receiver = Ndef.all['localhost']
	Meta_PrimitiveFailedError:new	0x113395840
		arg this = PrimitiveFailedError
		arg receiver = Ndef.all['localhost']
	Object:primitiveFailed	0x112181fc0
		arg this = Ndef.all['localhost']
	String:compare	0x11439db80
		arg this = Ndef.all['localhost']
		arg aString = nil
		arg ignoreCase = false
	String:!=	0x11439f5c0
		arg this = Ndef.all['localhost']
		arg aString = nil
	ProxyMixer:checkUpdate	0x11327f900
		arg this = a NdefMixer
		var newState = ( 'object': ProxySpace ()
, 'name': Ndef.all['localhost'], 'arNames': [  ], 'krNames': [  ], 
  'arOverflow': 0, 'krOverflow': 0 )
		var arNames = nil
		var prevArNames = nil
		var fullSize = nil
		var krNames = nil
		var prevKrNames = nil
		var editName = nil
		var prevEditName = nil
	a FunctionDef	0x113c4c740
		sourceCode = "<an open Function>"
	Routine:prStart	0x113b92bc0
		arg this = a Routine
		arg inval = 16.911234227

CALL STACK:
	MethodError:reportError   0x11210f1b8
		arg this = <instance of PrimitiveFailedError>
	< closed FunctionDef >   0x112164068
		arg error = <instance of PrimitiveFailedError>
	Integer:forBy   0x115003c58
		arg this = 0
		arg endval = 2
		arg stepval = 2
		arg function = <instance of Function>
		var i = 2
		var j = 1
	SequenceableCollection:pairsDo   0x112057258
		arg this = [*4]
		arg function = <instance of Function>
	Scheduler:seconds_   0x11205a488
		arg this = <instance of Scheduler>
		arg newSeconds = 17.487860289
	Meta_AppClock:tick   0x11205cbb8
		arg this = <instance of Meta_AppClock>
		var saveClock = <instance of Meta_SystemClock>
	Process:tick   0x112059828
		arg this = <instance of Main>
^^ The preceding error dump is for ERROR: Primitive '_StringCompare' failed.
Wrong type.
RECEIVER: Ndef.all['localhost']
```

Could you take a look? One solution would be to allow Symbols more broadly in methods which take Strings as arguments. This would help with efficiency in many cases, as it would avoid creation of objects which are used once and then discarded, which the need to convert makes necessary. That may need deeper consideration though.

IAC, would be good to merge this PR soon, and yes before 3.7.

**Comment by jamshark70** [comment]
Adding support for string comparison against symbols is a valid idea (though not currently supported!), but that's irrelevant to these stack traces, which are for someString != nil.

For <, <=, > and >=, it should throw an error to compare a string against something not string-like. However, for == and !=, it's reasonable to return false and true respectively if the classes don't match (and, in those cases, don't call 'compare' at all).

This is probably why JMc elected not to return an error from the string compare primitive: so that these two cases "just work."

**Comment by jamshark70** [comment]
Sorry, fat finger on the phone.

It's better to throw a user-readable error in the faulty cases, instead of the current situation where < etc. appear not to be understood.

**Comment by telephon** [comment]
OK, @muellmusik, could you perhaps add this to String.sc:

```
== { arg aString; if(aString.isString.not) { ^false }; ^this.compare(aString, false) == 0 }
!= { arg aString; if(aString.isString) { ^false }; ^this.compare(aString, false) != 0 }
```

**Comment by telephon** [comment]
>IAC, would be good to merge this PR soon, and yes before 3.7.

yes, we should do this. But just briefly reconsider:

```[5, \tte, 71, "hello", "aago"].sort```This used to work. Does it make sense to not allow it or vice versa?

**Comment by telephon** [comment]
>One solution would be to allow Symbols more broadly in methods which take Strings as arguments. This would help with efficiency in many cases, as it would avoid creation of objects which are used once and then discarded, which the need to convert makes necessary. That may need deeper consideration though.

This would be very good indeed. Also it would be good to add support for ````String```` in OSC, so that we don't fill up the symbol table while we are chatting along over the net in sclang.

It would need deeper consideration because symbols are used in specific ways as flags neutral to math operations that can be used next to them. They are in a sense like a neutral element of abstract functions (a bit like ```x = { thisFunction }```).

**Comment by jamshark70** [comment]
!= { arg aString; if(aString.isString) { ^false }; ^this.compare(aString, false) != 0 }

The "if" for != should return true.

**Comment by telephon** [comment]
Ah yes. corrected above.

**Comment by muellmusik** [comment]
> Adding support for string comparison against symbols is a valid idea (though not currently supported!), but that's irrelevant to these stack traces, which are for someString != nil.

Ah sorry @jamshark70 and @telephon. Just threw up a quick dump in a rush cause I had a problem in rehearsal. Should know better.

> This is probably why JMc elected not to return an error from the string compare primitive: so that these two cases "just work."

Don't think so, based on some research I did. ;-) IAC, I'd think that should be implemented in String as Julian suggests.

**Comment by muellmusik** [comment]
> IAC, would be good to merge this PR soon, and yes before 3.7.
> yes, we should do this. But just briefly reconsider:
> 
> [5, \tte, 71, "hello", "aago"].sortThis used to work. Does it make sense to not allow it or vice versa?
> 
> But this doesn't:
> 
> [5, \tte, 71, "hello", "aago", [1, 2, 3]].sort
> 
> Should we allow comparison between partially orderable objects by providing a specific algorithm (as it is done in string)?

I'm not sure. If yes and yes, let's make sure this is defined and documented clearly. String:compare says "Returns a -1, 0, or 1 depending on whether the receiver should be sorted before the argument, is equal to the argument or should be sorted after the argument." But this is not true:
```
"foo".compare("bar")
-> 4
```

**Comment by telephon** [comment]
The implementation uses ```memcmp``` – http://www.cplusplus.com/reference/cstring/memcmp/

**Comment by telephon** [comment]
… so unless we have a really consistent way of comparing objects of arbitrary type, we should rather throw an error.

**Comment by crucialfelix** [comment]
What's the status of this ?

**Comment by muellmusik** [comment]
@crucialfelix Needs a decision about the Symbol issue. But should definitely be merged for 3.7.

**Comment by crucialfelix** [comment]
OK, can we get it ready for merge soon ?  thanks.

**Comment by telephon** [comment]
I've had this patch running for two weeks or so, and haven't run ito problems.

**Comment by muellmusik** [comment]
@telephon Do you have a preference for how to proceed on the Symbol issue? NdefMixer is the only place I've seen a problem, so one approach would be to fix it there.

**Comment by telephon** [comment]
if nil is involved in a comparison with a string, ```==``` and ```!=``` should return ```false``` and ```true``` respectively, no? Orderings should throw an error, I think.

**Comment by telephon** [comment]
In other words, the following should not throw an error:

```\x == nil```

```\x != nil```

But this should:

```\x > nil```

**Comment by muellmusik** [comment]
Yes.

**Comment by muellmusik** [comment]
Or actually I'm not sure. It seems the current behaviour is intentional.

```
case tagSym :
			if (IsSym(b)) {
				switch (opcode) {
					case opEQ  : SetBool(a, slotRawSymbol(a) == slotRawSymbol(b)); break;
					case opNE  : SetBool(a, slotRawSymbol(a) != slotRawSymbol(b)); break;
					case opLT  : SetBool(a, strcmp(slotRawSymbol(a)->name, slotRawSymbol(b)->name) <  0); break;
					case opGT  : SetBool(a, strcmp(slotRawSymbol(a)->name, slotRawSymbol(b)->name) >  0); break;
					case opLE  : SetBool(a, strcmp(slotRawSymbol(a)->name, slotRawSymbol(b)->name) <= 0); break;
					case opGE  : SetBool(a, strcmp(slotRawSymbol(a)->name, slotRawSymbol(b)->name) >= 0); break;
					//default : leave first operand on stack
				}
			} else {
				if (IS_BINARY_BOOL_OP(opcode))
					SetBool(a, opcode == opNE);
			}
			break;
```
Symbols are compared using strcmp. Otherwise return false unless op is !=

So perhaps it's okay as in this PR.

**Comment by muellmusik** [comment]
So I think I've lost the plot on this. What do we need to do to merge this? Just fix the problem in NdefMixer? Revert the String:compare commit?

**Comment by telephon** [comment]
>Or actually I'm not sure. It seems the current behaviour is intentional.

What about this: merge everything without the string compare modification, and add another branch for that so we can carefully look into it for 3.7.1?

**Comment by muellmusik** [comment]
> What about this: merge everything without the string compare modification, and add another branch for that so we can carefully look into it for 3.7.1?

Well, I think we *might* reasonably think that the behaviour is now correct for the examples you give.

But this ```"foo" > 5``` throws an error saying that nil doesn't understand >. Huh?

**Comment by adcxyz** [comment]
Hi, 
while it is easy to fix the test in ProxyMixer for the correct stricter primitive behavior, 
testing "astring" == nil or also nil == "str" should just return false, not throw an error. 
It seems reasonable to expect one can test any two objects for non/equality 
without ever seeing an error - compareObject does this nicely.

So for strings, I think the implementations should follow what compareObject does
(which expands a bit on what JH suggested):  
== { arg aString; 
	if (aString === this) { ^true }; 
	if (aString.isString.not) { ^false }; 
	^this.compare(aString, false) == 0 
}
!= { arg aString; 
	if (aString === this) { ^false }; 
	if (aString.isString.not) { ^true }; 
	^this.compare(aString, false) != 0 
}

BTW, briefly looking at some implementations of ==, the one for Association is wrong, 
as it never compares the values: 
	== { arg anAssociation;
		^anAssociation respondsTo: \key and: { key == anAssociation.key }
	}
(\a -> 123) == (\a -> "foo"); // true!

... and the one for Warp as well: 
a = [1, 10, \lin].asSpec.warp;
b = [1, 100, \lin].asSpec.warp;
a == b; // true

There are 32 implementations of ==. 
Would be worth going through all of them, no?

**Comment by muellmusik** [comment]
> But this "foo" > 5 throws an error saying that nil doesn't understand >. Huh?

Actually must have been running an earlier build. This branch *fixes* that with a sensible error

**Comment by muellmusik** [comment]
Okay, so just to sum up the issues, and try to reach agreement:

There seems general agreement that aStringOrSymbol == or != someOtherObject should never throw an error, yes?

>, >=, <, <= are more complicated, but should possibly be consistent for Strings and Symbols? Compare:

In Symbols these always return false unless the second object is also a Symbol, in which case it uses strcmp.

In Strings it uses memcmp to compare. As of this branch, it throws an error if the other object's not a String. Prior to this branch it would compare the receiver with whatever was in u of the other slot (i.e. ```memcmp(slotRawString(a)->s, slotRawString(b)->s, length);```). This strikes me as unlikely to produce reliably sensible results, considering all the possibilities that could be in u. (Maybe there's something clever here, but if so the implementation is a little opaque.)

For further info, >, >=, <, <= are undocumented for Symbol (though implemented), and for String documented as used for sorting the receiver relative to another String.

My feeling is that the current String behaviour is anyway wrong, and should be made the same as Symbol. (We could go the other way, but consistent in any case). Comparing Strings to Symbols and vice versa is a new feature and could wait until 3.8.

Agreed?

**Comment by telephon** [comment]
Agreed.

**Comment by muellmusik** [comment]
> Agreed

Okay, that is easily done. One other thing I noticed though, that ```5 > "foo"``` eventually takes us to the following, which converts 5 to a String.
```
	performBinaryOpOnSimpleNumber { arg aSelector, aNumber;
		^aNumber.asString.perform(aSelector, this);
	}
```
So numbers are automagically converted to Strings and compared. This is undocumented behaviour. 

@jamshark70 I'm afraid this was from 0ffa1cb. Is there a reason why this was deemed better than to require explicit conversions? (That seems safest to me.) Also why only one way? (i.e. what's the correct reverse of this, convert the String to a number if possible, or?) Looking at the way it was before 0ffa1cb it seems possible to me that the intention was that it would throw an error?

**Comment by muellmusik** [comment]
My instinct here is that conversions for comparison between similar objects (e.g. Int->Float, perhaps String->Symbol) are uncontroversial? But more exotic ones (number->String) are more of an issue?

**Comment by telephon** [comment]
>My instinct here is that conversions for comparison between similar objects (e.g. Int->Float, perhaps String->Symbol) are uncontroversial? 

yes.

>But more exotic ones (number->String) are more of an issue?

The question is effectively if you want to be able to sort a heterogenous array of things in a consistent way. The last resort would be sorting by hash. Maybe we need to implement comparison properly and completely for all objects, and then fall back on hash when nothing is defined. A primitive can only be a small part of that, I think.

**Comment by muellmusik** [comment]
> The question is effectively if you want to be able to sort a heterogenous array of things in a consistent way. The last resort would be sorting by hash. Maybe we need to implement comparison properly and completely for all objects, and then fall back on hash when nothing is defined. A primitive can only be a small part of that, I think.

Okay, so should I fix as agreed above, and leave this and the number case for post 3.7?

**Comment by telephon** [comment]
Yes, that would be good.

**Comment by muellmusik** [comment]
The travis failure seems to be the fail to deploy problem.

**Comment by jamshark70** [comment]
> @jamshark70 I'm afraid this was from 0ffa1cb. Is there a reason why this was deemed better than to require explicit conversions? (That seems safest to me.) Also why only one way? (i.e. what's the correct reverse of this, convert the String to a number if possible, or?) Looking at the way it was before 0ffa1cb it seems possible to me that the intention was that it would throw an error?

That was some 7 years ago and, in the absence of a text description in the commit message, there's really no way I can remember.

I don't object to changing it so that it throws a meaningful error.

**Comment by muellmusik** [comment]
Merge?

**Pull request merged**